Signaling - Gravity bridge deployment (#27)
cosmos/governance: proposals/2021-05-gravity-bridge-deployment/README.md

# Signaling Proposal - Deployment of Gravity Bridge on the Cosmos Hub

![Gravity Bridge](https://raw.githubusercontent.com/althea-net/cosmos-gravity-bridge/main/gravity-bridge.svg)

## Summary

This proposal is a Request For Comment to the ATOM community regarding the activation of the Gravity Bridge module onto the Cosmos Hub.

By voting YES to this proposal, you will signal that you approve of having the Gravity Ethereum <> Cosmos bridge deployed onto the Cosmos Hub.

## Vision

> Gravity as an Ethereum-Cosmos bridge is designed for the Cosmos Hub, to pull as much value as possible into the orbits of Cosmos via a direct and decentralized bridge. Gravity will be able to bring ERC20 assets from Ethereum into Cosmos, as well as Cosmos assets to Ethereum ERC20 representations. ATOM, and any other asset in the Cosmos ecosystem, will be able to trade on Uniswap and other Ethereum AMMs, and interact with Ethereum DeFi like any ERC20 token. This will bring a tremendous amount of liquidity and utility to these assets.

Reviewer: ```suggestion
Gravity as an Ethereum-Cosmos bridge is designed for the Cosmos Hub to pull as much value as possible into the orbits of Cosmos via a direct and decentralized bridge. Gravity will be able to bring ERC20 assets from Ethereum into Cosmos, as well as Cosmos assets to Ethereum ERC20 representations. ATOM, and any other asset in the Cosmos ecosystem, will be able to trade on Uniswap and other Ethereum AMMs, and interact with Ethereum DeFi like any ERC20 token. This will bring a tremendous amount of liquidity and utility to these assets.
```

> Gravity as an Ethereum-Cosmos bridge is designed for the Cosmos Hub, to pull as much value as possible into the orbits of Cosmos via a direct and decentralized bridge. Gravity will be able to bring ERC20 assets from Ethereum into Cosmos, as well as Cosmos assets to Ethereum ERC20 representations. ATOM, and any other asset in the Cosmos ecosystem, will be able to trade on Uniswap and other Ethereum AMMs, and interact with Ethereum DeFi like any ERC20 token. This will bring a tremendous amount of liquidity and utility to these assets.

Reviewer: ```suggestion
Gravity as an Ethereum-Cosmos bridge is designed for the Cosmos Hub to pull as much value as possible into the orbits of Cosmos via a direct and decentralized bridge. Gravity will be able to bring ERC20 assets from Ethereum into Cosmos, as well as Cosmos assets to Ethereum ERC20 representations. ATOM, and any other asset in the Cosmos ecosystem, will be able to trade on Uniswap and other Ethereum AMMs, and interact with Ethereum DeFi like any ERC20 token. As well, Cosmos Hub users can use the Gravity DEX to trade between ERC20 assets and tokens that are transferred using IBC. This will bring a tremendous amount of liquidity and utility to these multi-chain assets.
```

## Cosmos, Ethereum, and Gravity

> Gravity is a secure and highly efficient bridge between EVM and Cosmos SDK-based blockchains. At a high-level, Gravity enables token transfers from Ethereum to the Cosmos Hub and back again, by locking up tokens on the Ethereum side and minting equivalent tokens on the Cosmos side.

Reviewer: ```suggestion
Gravity is a secure and highly efficient bridge between EVM and Cosmos SDK-based blockchains. At a high-level, Gravity enables token transfers from Ethereum to the Cosmos Hub and back again by locking tokens on the Ethereum side and minting equivalent tokens on the Cosmos side.
```

Gravity is completely non-custodial. Control of the bridge mirrors the active validator set on the Cosmos SDK-based chain, and validator stake on Cosmos can be slashed for misbehavior involving the Gravity bridge.

The Gravity Ethereum contract is highly optimized, utilizing batches to dramatically reduce the cost of Cosmos -> Ethereum transfers. Sending funds from Cosmos back to Ethereum can be up to 50% less costly than a normal ERC20 send.

## How do validators support the Gravity Bridge?

Cosmos Hub validators will run three key software components of the Gravity bridge:

> - The Gravity bridge module, integrated into gaiad (the main Cosmos binary)

Reviewer: ```suggestion
- The Gravity bridge module, integrated into gaiad (the binary that runs the Cosmos Hub)
```

- The Gravity bridge Orchestrator
- A Geth light client or any Ethereum full node implementing the JSON-rpc standard

### Cosmos to Ethereum:

> To send transactions from Cosmos to Ethereum, the Gravity Bridge module first packages the transaction data, and makes it available on an endpoint. The Orchestrator then signs this data with the validator’s Ethereum key, and submits it as a message. These signatures will then be assembled and submitted to the Ethereum chain by relayers.

Reviewer: ```suggestion
To send transactions from Cosmos to Ethereum, the Gravity Bridge module of a validator's Gaia instance first packages the transaction data, and makes it available on an endpoint. The Orchestrator then signs this data with the validator’s Ethereum key, and submits it as a message to the Ethereum network. The Ethereum signature of each Cosmos validator needs to be assembled and submitted to the Ethereum chain by relayers.
```

> Validators may be slashed if they fail to submit Ethereum signatures within 6000 blocks (about eight hours) of their creation.

Reviewer: ```suggestion
Validators may be slashed if they fail to submit Ethereum signatures within 6000 Cosmos Hub blocks (about eight hours) of their creation.
```

> Validators may be slashed if they fail to submit Ethereum signatures within 6000 Cosmos Hub blocks (about eight hours) of their creation.

Reviewer 2: Would be good to mention this in comparison to the existing liveness slashing rule (ie. failing to sign for 9500 blocks, or more accurately failing to sign at least 500 of the last 10,000 blocks):

```
$ gaiad query slashing params
min_signed_per_window: "0.050000000000000000"
signed_blocks_window: "10000"
```

Author: added a comparison

Validators may also be slashed if they sign a message with their Ethereum key that was not created by the Gravity bridge module.

Gravity bridge has no other slashing conditions.

### Ethereum to Cosmos:

> The Orchestrator also monitors the Ethereum chain, submitting events that occur on Ethereum to Cosmos as messages. When more than 66% of the active voting power has sent a message observing the same Ethereum event, the Gravity module will take action.

Reviewer: ```suggestion
The Orchestrator also monitors the Ethereum chain, submitting events that occur on Ethereum to Cosmos as messages. When more than 66% of the active Cosmos voting power has sent a message observing the same Ethereum event, the Gravity module will take action.
```

This oracle action will not be incentivized, nor will it be enforced with slashing. If validators making up more than 33% of the staked tokens do not participate in the oracle, new deposits and withdrawals will not be processed until those validators resume their oracle obligations.

> The oracle was designed without slashing conditions to ensure that a consensus failure on Ethereum does not affect operation of the Cosmos chain.

Reviewer: ```suggestion
The oracle was designed to adhere to the Cosmos Hub validator security model without slashing conditions to ensure that a consensus failure on Ethereum does not affect operation of the Cosmos chain.
```

### Slashing Conditions Spec

https://github.com/cosmos/gravity-bridge/blob/main/spec/slashing-spec.md

## How does it work?

Gravity consists of 4 parts:

- An Ethereum contract called Gravity.sol

> - The Gravity Cosmos module

Reviewer: ```suggestion
- The Gravity Cosmos module of Gaia
```

- The orchestrator program which is run by Cosmos validators alongside the Gravity module
- A market of relayers, who compete to submit transactions to Ethereum

### Gravity.sol

> The Gravity Ethereum contract is a highly compact and efficient representation of weighted powers voting on Ethereum. It contains an Etheruem key from each Cosmos validator, as well as their voting power. This signer set is continuously updated as validation power changes on Cosmos, ensuring that it matches the current Cosmos validator set.

Reviewer 3: ```suggestion
The Gravity Ethereum contract is a highly compact and efficient representation of weighted powers voting on Ethereum. It contains an Ethereum key from each Cosmos validator, as well as their voting power. This signer set is continuously updated as validation power changes on Cosmos, ensuring that it matches the current Cosmos validator set.
```

> Sending tokens, or updating the validator set, contained in Gravity.sol requires more than 66% of the total voting power to approve the action. In this way Gravity.sol mirrors Tendermint consensus on the Cosmos chain as closely as possible on Ethereum.

Reviewer 2: Minor nit but presumably all reference to 66% is actual 2/3 (ie 0.6666666...) ? We've traditionally just referred to this as `+2/3 of voting power`

Author: I've switched it to the proper fraction.

### Gravity Cosmos Module

The Gravity module governs and coordinates the bridge. Generating messages for the validators to sign with their Ethereum keys and providing these signatures to relayers who assemble and submit them to the Ethereum chain.

### Orchestrator

> The Gravity bridge orchestrator performs all the external tasks the Gravity bridge requires for validators. The submission of signatures and the submission of Ethereum events.

Reviewer: ```suggestion
The Gravity bridge orchestrator performs all the external tasks the Gravity bridge requires for validators, which includes submission of signatures and submission of Ethereum events.
```

> While the Gravity module concerns itself with the correctness and consensus state of the bridge the Orchestrator locates or creates the correct inputs.

Reviewer: ```suggestion
While the Gravity module concerns itself with the correctness and consensus state of the bridge, the Orchestrator is responsible for locating and creating the correct inputs.
```

### Market of Relayers

Relayers are an unpermissioned role that observes the Cosmos chain for messages ready to be submitted to Ethereum.

> The relayer then packages the validators signatures into an Ethereum transaction and submits that transaction to the Ethereum blockchain. All rewards in the Gravity bridge design are paid to msg.sender on Ethereum. This means that relayers do not require any balance on the Cosmos side and can immediately liquidate their earnings into ETH to relay future messages.

Reviewer: ```suggestion
The relayer then packages the validators signatures into an Ethereum transaction and submits that transaction to the Ethereum blockchain. All rewards in the Gravity bridge design are paid to msg.sender on Ethereum. This means that relayers do not require any balance on the Cosmos side and can immediately liquidate their earnings into ETH while continuing to relay newer messages.
```

## Security assumptions

> The Gravity bridge is designed with the assumption that the total amount of funds in Gravity.sol is less than the value of the validator set’s total staked tokens.

Reviewer 2: Doesn't this assume 100% slashing? Right now slashing on the Hub is parameterized at 5% for equivocation and 0.01% for downtime:

```
$ gaiad query slashing params
slash_fraction_double_sign: "0.050000000000000000"
slash_fraction_downtime: "0.000100000000000000"
```

Will Gravity expose its own set of parameters here ?

Author: yes gravity will expose it's own set of parameters. I need to clarify some issues here as well.

Author: https://github.com/althea-net/cosmos-gravity-bridge/blob/main/spec/slashing-spec.md#gravslash-03-failure-to-sign-validator-set-update

Please see the second on `UnbondAttackSlashing` to cover this case.

If this assumption does not hold true, it would be more profitable for validators to steal the funds in the bridge and simply lose their stake to slashing.

> There is no automated enforcement of this assumption. It is up to the $ATOM holders to take action if the amount deposited in the bridge exceeds the total value of all stake on the hub.

Reviewer 3: Can we be a bit more precise regarding "taking action" here?

Author: Validators could either ensure more tokens enter the staking pool or take action to move what assets they control back across the bridge to reduce the total value in the bridge.

An option to allow the validators to disable deposits might be warranted, but would be complex to implement at this stage.

> It should be noted that this condition is not unique to the Gravity bridge. The same dynamic exists for any IBC connection, and even exists in scenarios other than cross-chain communication. For example, in a hypothetical blockchain keeping domain name records, this same vulnerability would exist if the potential profit from exploiting the domain name system was greater than the value of the validator set’s total staked tokens.

Reviewer 2: It would be great to include some description of how Gravity compares to IBC and why it was not possible or otherwise ideal to use IBC directly

Author: I've added a section on why we can not use IBC to bridge to Ethereum, a section comparing IBC and Gravity was previously attempted but removed due to the amount of technical nuance. I'd be interested in what properties you think are the most important to compare?

Interchain Foundation

All in Bits/Tendermint

Sommelier, Informal, Injective, Confio

_Gravity Readiness Committee:_

Justin Kilpatrick and Jehan Tremback, Althea

Zarko Milosevic, Informal Systems

Zaki Manian, Sommelier/Iqlusion

## Governance Votes

The following items summarize the voting options and what it means for this proposal.

- **YES**: You agree that Gravity Bridge should be deployed to the Cosmos Hub.
- **NO**: You disapprove of deploying Gravity bridge on the Cosmos Hub in its current form (please indicate in the [Cosmos Forum](https://forum.cosmos.network/) why this is the case).
- **NO WITH VETO**: You are strongly opposed to the deployment of Gravity bridge on the Cosmos Hub and will exit the network if this occurred.
- **ABSTAIN**: You are impartial to the outcome of the proposal.

## Appendix

### FAQ

### Is running the Gravity Module difficult for Cosmos Validators?

Soliciting feedback from dozens of Cosmos hub validators and over 100 test net participants, we found that running the Gravity module is not a difficult task or undue burden on Cosmos validators.

### Is the Gravity bridge secure?

> The Gravity bridge is undergoing an audit by Informal Systems. It will then be up to ATOM holders to interpret the results of the code audit and weigh implementation risks in another governance proposal before deployment. Fundamentally, the design of the Gravity Bridge means that its security is directly represented by the security of the validator set on Cosmos Hub.

Reviewer: ```suggestion
The Gravity bridge is undergoing an audit by Informal Systems. It will then be up to ATOM holders to interpret the results of the code audit and weigh implementation risks in another governance proposal before deployment. Fundamentally, the design of the Gravity Bridge means that its security is directly represented by the security of the validator set on Cosmos Hub. On the Cosmos side, the Gravity Bridge security design adheres to the Cosmos network's ideal of safety over liveness assured by 2/3+ validator voting power, but without sacrificing safety or liveness due to issues that may rarely arise with Ethereum.
```

### Are slashing conditions a problem for validators?

Gravity Bridge slashing conditions closely mirror the slashing conditions which validators are already subject to.

> - Uptime: Validators on Cosmos currently must keep their validator software running at all times, or risk slashing. Gravity adds an additional binary which must be run, which is low in difficulty.

Reviewer: ```suggestion
- Uptime: Validators on Cosmos currently must keep their validator software running at all times, or risk slashing. Gravity adds an additional binary which must be run, which is low in difficulty, resource usage, and operational requirements.
```

- Equivocation: Validators on Cosmos are subject to slashing if they sign two blocks at the same height. It is possible for this to happen through accidental misconfiguration. Gravity adds an additional item which must not be signed, which are the fraudulent bridge transactions that never existed on Cosmos. It is not possible for this to happen by accident, so this slashing condition is much less of a risk than the Hub’s existing slashing conditions.

> ### How is Gravity Bridge different from Peggy?
> 
> The original vision for Peggy outlined in the Cosmos whitepaper is a ‘peg-zone’, meaning a Cosmos blockchain that exists for the sole purpose of running the bridge to Ethereum. All tokens would then be moved to other more general purpose Cosmos chains via IBC.
> 
> This design is heavily reliant on IBC, not only for moving Ethereum tokens to where they will be used, but also to handle the stake for bridge operations. The validator set operating the bridge must have at least as much at stake as the value in the bridge. In the Peggy concept this would require some shared security system built on top of IBC.
> 
> Gravity is a simplified design for a bridge that relies directly on the stake of the Cosmos chain running the bridge. Bridges like IBC and Gravity are complex on their own, so by taking this strategy IBC and Gravity remain as distinct bridges operating on the same chain, rather than being tightly connected systems.
> 
> This is easier to both design and debug, and is ideal for high-value chains like the Hub.

Reviewer 4: ```suggestion
### What about peg zones?

The concept of a "peg zone" has been around in Cosmos for a while. This is a separate chain which runs the Gravity Bridge and connects to Ethereum or another blockchain. We believe that running Gravity Bridge on the Cosmos Hub and connecting directly is the superior solution. First, let's look at what different forms peg zones could take.

1. The most likely type of peg zone is what will result if this proposal does not pass. There are at least 5 Cosmos SDK chains who will be using the Gravity Bridge module to connect to Ethereum. There will be no official way to bridge Ethereum assets into the Cosmos ecosystem. Instead there will be many dueling representations of Ethereum assets. It goes without saying that this will confuse Cosmos users. If one of these peg zones gains the upper hand and becomes dominant, the Cosmos Hub will miss out on all the transaction fees that it generates.

2. It would also be possible to establish an official peg zone, and airdrop its staking token 1:1 to current Atom holders. This would at least allow the Cosmos Hub stakeholders to keep the economic benefit of activity on the peg zone. However, this is capital inefficient. A dollar staked  on this peg zone would not be staked on the hub and vice versa. Splitting stake between these two important chains would make both weaker. It would also confuse prospective Atom investors.

3. Shared security has been talked about for a long time in Cosmos, but it is not yet in production. Establishing a peg zone with shared security with the Cosmos Hub would allow the same validator set to validate both chains, and put the same Atoms at stake to secure each of them. This would avoid the issues of scenarios 1 and 2 above. However, we cannot afford to wait. Many other PoS blockchains already have Ethereum bridges and Cosmos needs to continue innovating to stay relevant.
```
I would switch this out for clarity, to explain all the relevant tradeoffs, and disambiguate between the old Peggy codebase (not relevant) and the concept of a peg zone (relevant)

Author: this is a good change, just waiting on those minor edits we discussed.
